fendermint/vm/interpreter/src/fvm/state/snapshot.rs

```diff
@@ -0,0 +1,357 @@
+    /// The list of cids to pull from the blockstore
+    #[pin]
+    dfs: VecDeque<Cid>,
+    /// The block store
+    bs: BlockStore,
+}
+
+impl<BlockStore> StateTreeStreamer<BlockStore> {
+    pub fn new(state_root_cid: Cid, bs: BlockStore) -> Self {
+        let mut dfs = VecDeque::new();
+        dfs.push_back(state_root_cid);
+        Self { dfs, bs }
+    }
+}
+
+impl<BlockStore: Blockstore> Stream for StateTreeStreamer<BlockStore> {
+    type Item = (Cid, Vec<u8>);
+
+    fn poll_next(self: Pin<&mut Self>, _cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        let mut this = self.project();
+
+        loop {
+            let cid = if let Some(cid) = this.dfs.pop_front() {
+                cid
+            } else {
+                return Poll::Ready(None);
+            };
+
+            match this.bs.get(&cid) {
+                Ok(Some(bytes)) => {
+                    let ipld = from_slice::<Ipld>(&bytes).unwrap();
+                    walk_ipld_cids(ipld, &mut this.dfs);
+                    return Poll::Ready(Some((cid, bytes)));
+                }
+                Ok(None) => {
+                    tracing::warn!("cid: {cid:?} has no value in block store, skip");
+                    continue;
+                }
+                Err(e) => {
+                    tracing::warn!("cannot get from block store: {}", e.to_string());
+                    // TODO: consider returning Result, but it won't work with `car.write_stream_async`.
+                    return Poll::Ready(None);
+                }
+            }
+        }
+    }
+}
+
+fn walk_ipld_cids(ipld: Ipld, dfs: &mut VecDeque<Cid>) {
+    match ipld {
+        Ipld::List(v) => {
+            for i in v {
+                walk_ipld_cids(i, dfs);
+            }
+        }
+        Ipld::Map(map) => {
+            for v in map.into_values() {
+                walk_ipld_cids(v, dfs);
+            }
+        }
+        Ipld::Link(cid) => dfs.push_back(cid),
+        _ => {}
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::fvm::state::snapshot::StateTreeStreamer;
+    use crate::fvm::state::{FvmStateParams, Snapshot};
+    use crate::fvm::store::memory::MemoryBlockstore;
+    use crate::fvm::store::ReadOnlyBlockstore;
+    use cid::Cid;
+    use fendermint_vm_core::Timestamp;
+    use futures_util::StreamExt;
+    use fvm::state_tree::{ActorState, StateTree};
+    use fvm_ipld_blockstore::Blockstore;
+    use fvm_shared::state::StateTreeVersion;
+    use fvm_shared::version::NetworkVersion;
+    use quickcheck::{Arbitrary, Gen};
+    use std::collections::VecDeque;
+
+    fn prepare_state_tree(items: u64) -> (Cid, StateTree<MemoryBlockstore>) {
+        let store = MemoryBlockstore::new();
+        let mut state_tree = StateTree::new(store, StateTreeVersion::V5).unwrap();
+        let mut gen = Gen::new(16);
+
+        for i in 1..=items {
+            let state = ActorState::arbitrary(&mut gen);
+            state_tree.set_actor(i, state);
```

Comment:
So the state will have a CID not pointing at anything, right?

Reply:
right, for test only. The cid will be picked up by the ipld walker and found there is no data there and continue. We can panic if this happens in non-test env.

Reply:
I wouldn't panic because CIDs without data were for example part of the atomic execution protocol at some point.

Reply:
I was only asking because I was wondering if that CID following is indeed exercised by this code as it is (the HAMT can store all data in a single node if there aren't too many elements), but it looks fine to me :crossed_fingers: